onnxmltools/utils/metadata_props.py

```diff
@@ -0,0 +1,67 @@
+from ..proto import onnx, onnx_proto
+from distutils.version import StrictVersion
+
+
+def _get_case_insensitive(iterable, key):
+    return next((x for x in iterable if x.lower() == key.lower()), None)
+
+
+def _validate_metadata(metadata_props):
+    valid_metadata_props = {
+        'Image.BitmapPixelFormat': ['Gray8', 'Rgb8', 'Bgr8', 'Rgba8', 'Bgra8'],
+        'Image.ColorSpaceGamma': ['Linear', 'SRGB'],
+        'Image.NominalPixelRange': ['NominalRange_0_255', 'Normalized_0_1', 'Normalized_1_1', 'NominalRange_16_235'],
+    }
+    image_metadata_props = {k: v for k, v in metadata_props.items() if _get_case_insensitive(valid_metadata_props, k)}
+    for key, value in image_metadata_props.items():
+        key = _get_case_insensitive(valid_metadata_props, key)
+        if not key:
+            print('Warning: key {} is defined multiple times'.format(key))
+        else:
+            valid_values = valid_metadata_props.pop(key)
+            if not _get_case_insensitive(valid_values, value):
+                print('Warning: value {} is invalid. Valid values are {}'.format(value, valid_values))
+    if image_metadata_props and valid_metadata_props:
+        print('Warning: incomplete image metadata is being added. Keys {} are missing.'.format(', '.join(valid_metadata_props)))
+
+COLOR_SPACE_TO_PIXEL_FORMAT = {
+    'BGR': 'Bgr8',
+    'RGB': 'Rgb8',
+    'GRAY': 'Gray8',
+}
+
+
+def color_space_to_pixel_format(color_space):
+    return COLOR_SPACE_TO_PIXEL_FORMAT[color_space]
+
+
+def add_metadata_props(onnx_model, metadata_props, targeted_onnx=onnx.__version__):
+    if StrictVersion(targeted_onnx) < StrictVersion('1.2.1'):
+        print('Metadata properties are not supported in targeted ONNX-%s' % targeted_onnx)
+        return
+    _validate_metadata(metadata_props)
+    # Overwrite old properties (case insensitive)
+    new_props = [x.lower() for x in metadata_props]
+    model_metadata = onnx_model.metadata_props
+    for prop in model_metadata:
+        if prop.key.lower() in new_props:
+            model_metadata.remove(prop)
```

Comment:
Why there is a possibility that we add something and then remove it? Could this be avoided?

Reply:
I don't get your point... Metadata props are serialized in the protobuf as a list of `StringStringEntryProto`. If the user calls the function to set some props, any existing entries with the same name have to be removed. We have to remove duplicate entries before extending the list.

Reply:
Well, I assume this function is an internal one so that user will never use it.